witchcraft/witchcraft.go

```diff
@@ -1035,10 +1050,27 @@ func decryptNodeValues(n *yamlv3.Node, kwt *encryptedconfigvalue.KeyWithType) er
 }
 
 func stopServer(s *Server, stopper func(s *http.Server) error) error {
-	if s.stateManager.State() == ServerIdle {
-		return werror.Error("server is not running")
+	// use compare and swap so that the server can only be stopped once
+	curState := s.stateManager.State()
+	for {
+		if curState == ServerIdle || curState == ServerShutdownInProgress {
+			// already shutting down or stopped
+			return nil
+		}
+		// state could be ServerRunning or ServerInitializing
+		if s.stateManager.compareAndSwapState(curState, ServerShutdownInProgress) {
+			break
+		}
+		curState = s.stateManager.State()
```

Comment:
What's the condition where this occurs (either of the above states are not met)? Trying to understand since in that scenario this will be a busy/active wait

Reply:
Two shutdown attempts happen at the same time. Both check the current state and see it as running (i.e. both read the state before either sets it). Then both set it to shutdown in progress and run the stopper. Now there is no single shutdown in progress, there are multiple.

Using atomics with compare and swap ensures that only one will win. This is a busy wait, but will almost never be run more than once, and its basically impossible to run more than twice. If you lost the first iteration, then you get the state again and see that a shutdown is already in progress so you let that thread handle it.